### ISSUE TITLE
"*ERROR*: No thread allocated" after open the bdev successfully

### ISSUE THREAD

**Issue by xxks-kkk** [spdk/spdk#538]
I'm working on developing Rust wrapper to SPDK and I'm testing my wrapper using `hello_bdev.c` example. I can successfully open the bdev using my rust wrapper to `spdk_bdev_open`. However, when I try to call the `spdk_bdev_get_io_channel` through my wrapper, the function get called with error:

```
thread.c: 442:spdk_get_io_channel: *ERROR*: No thread allocated
```

I'm wondering what's the possible reason for such error?

## Current Behavior
see above. 

## Steps to Reproduce

[Here](https://github.com/utsaslab/rustfs/blob/b2365d7f9709b6f009423e9492275089ff579012/examples/hello_nvme_bdev_rust_wrapper/src/main.rs#L66) is the line to `spdk_bdev_get_io_channel` call. My wrapper to the call is [here](https://github.com/utsaslab/rustfs/blob/b2365d7f9709b6f009423e9492275089ff579012/spdk-rs/src/bdev.rs#L119). I'm not sure whether this has something to do on how I use `spdk_app_start`.

Currently, I pass in rust async I/O framework main function to the `spdk_app_start` ([here](https://github.com/utsaslab/rustfs/blob/b2365d7f9709b6f009423e9492275089ff579012/examples/hello_nvme_bdev_rust_wrapper/src/main.rs#L83)). The async I/O is based on future and it is executed, I believe, in the same thread as `spdk_app_start`. I'm not sure whether executing in the same thread will cause the error?

## Context (Environment including OS version, SPDK version, etc.)

- SPDK 17.10
- Ubuntu 16.04
- Working in QEMU with emulated NVMe SSD

**Comment by jimharris** [comment]
Hi @xxks-kkk,

You'll need to call spdk_allocate_thread().  This allocates the contexts needed before creating a per-thread IO channel for the specified bdev.

You may want to check the bdev fio_plugin for examples on how to use spdk_allocate_thread().  The bdev fio plugin was purposely designed to *not* rely on the SPDK event framework, so it shows how to plumb the start/stop poller and send message functionality.

-Jim

**Comment by xxks-kkk** [comment]
Hello @jimharris 

Thanks for the quick response. I'm a little confused with explicitly calling `spdk_allocate_thread`. Currently, I still use `spdk_app_start` (i.e., the event framework) as the entry point of my program. According to `hello_bdev` program, `spdk_app_start` expects a callback function and the whole event framework will terminate if some following callback function calls `spdk_app_stop`.

For my setup, I pass in the entry point of Rust Async I/O framework (e.g., `tokio::run_async(run())`) to the `spdk_app_start`. The Rust Async I/O framework contains a reactor and pollers that will eventually terminate when all the futures are realized. At that point, `spdk_app_stop` will be called.

From these two setups, I couldn't see the reasoning why I have to explicitly allocate a thread. The only difference between the original `hello_bdev` and my current setup is that `hello_bdev` uses a chain of callbacks while now I use an async framework. My current implementation still relies on SPDK event framework.

Thanks!

**Comment by jimharris** [comment]
Hi @xxks-kkk,

Thanks for the extra details.  Without seeing the code, I can't say for sure what's happening.  I would suggest adding some instrumentation in lib/thread/thread.c to try to track this down.  You'll want to print out pthread_t values (using pthread_self()) in spdk_allocate_thread, spdk_free_thread and spdk_get_thread.  It might be a lot of output, but will help us know for sure where to look next.

-Jim

**Comment by xxks-kkk** [comment]
Hello @jimharris,

Thank for the follow-up. I added `pthread_self` to the functions you mentioned immediately after they acquire the lock. For example, in `spdk_get_thread`:

```
pthread_mutex_lock(&g_devlist_mutex);

SPDK_NOTICELOG("pthread_self: %d\n", pthread_self());

thread = _get_thread();
```

The print out looks like below:

```
Enter main
Starting SPDK v18.07 / DPDK 18.05.0 initialization...
[ DPDK EAL parameters: hello_bdev -c 0x1 --legacy-mem --file-prefix=spdk_pid19393 ]
EAL: Detected 4 lcore(s)
EAL: Detected 1 NUMA nodes
EAL: Multi-process socket /var/run/dpdk/spdk_pid19393/mp_socket
EAL: No free hugepages reported in hugepages-1048576kB
EAL: Probing VFIO support...
EAL: VFIO support initialized
app.c: 530:spdk_app_start: *NOTICE*: Total cores available: 1
reactor.c: 718:spdk_reactors_init: *NOTICE*: Occupied cpu socket mask is 0x1
thread.c: 120:spdk_allocate_thread: *NOTICE*: pthread_self: 460651776
reactor.c: 492:_spdk_reactor_run: *NOTICE*: Reactor started on core 0 on socket 0
thread.c: 197:spdk_get_thread: *NOTICE*: pthread_self: 460651776
vbdev_passthru.c: 374:vbdev_passthru_init: *NOTICE*: conf parse matched: Malloc1
thread.c: 197:spdk_get_thread: *NOTICE*: pthread_self: 460651776
vbdev_passthru.c: 478:vbdev_passthru_register: *NOTICE*: Match on Malloc1
vbdev_passthru.c: 512:vbdev_passthru_register: *NOTICE*: io_device created at: 0x0x5619e99a8ed0
vbdev_passthru.c: 523:vbdev_passthru_register: *NOTICE*: bdev opened
vbdev_passthru.c: 534:vbdev_passthru_register: *NOTICE*: bdev claimed
thread.c: 197:spdk_get_thread: *NOTICE*: pthread_self: 460651776
vbdev_passthru.c: 545:vbdev_passthru_register: *NOTICE*: pt_bdev registered
vbdev_passthru.c: 546:vbdev_passthru_register: *NOTICE*: created pt_bdev for: PT0
thread.c: 197:spdk_get_thread: *NOTICE*: pthread_self: 460651776
EAL: PCI device 0000:02:00.0 on NUMA socket 0
EAL:   probe driver: 8086:953 spdk_nvme
EAL:   0000:02:00.0 VFIO group is not viable!
EAL: Requested device 0000:02:00.0 cannot be used
bdev_nvme.c:1388:bdev_nvme_library_init: *ERROR*: NVMe SSD "0000:00:04.0" could not be found.
bdev_nvme.c:1389:bdev_nvme_library_init: *ERROR*: Check PCIe BDF and that it is attached to UIO/VFIO driver.
bdev_nvme.c:1388:bdev_nvme_library_init: *ERROR*: NVMe SSD "0000:02:00.0" could not be found.
bdev_nvme.c:1389:bdev_nvme_library_init: *ERROR*: Check PCIe BDF and that it is attached to UIO/VFIO driver.
thread.c: 197:spdk_get_thread: *NOTICE*: pthread_self: 460651776
thread.c: 197:spdk_get_thread: *NOTICE*: pthread_self: 460651776
thread.c: 197:spdk_get_thread: *NOTICE*: pthread_self: 460651776
thread.c: 197:spdk_get_thread: *NOTICE*: pthread_self: 460651776
thread.c: 197:spdk_get_thread: *NOTICE*: pthread_self: 460651776
thread.c: 197:spdk_get_thread: *NOTICE*: pthread_self: 460651776
thread.c: 197:spdk_get_thread: *NOTICE*: pthread_self: 460651776
thread.c: 197:spdk_get_thread: *NOTICE*: pthread_self: 460651776
thread.c: 197:spdk_get_thread: *NOTICE*: pthread_self: 460651776
bdev name: Malloc0
bdev name: Malloc1
bdev name: PT0
Successfully open the device
thread.c: 448:spdk_get_io_channel: *ERROR*: No thread allocated
Successful
thread.c: 197:spdk_get_thread: *NOTICE*: pthread_self: 460651776
thread.c: 197:spdk_get_thread: *NOTICE*: pthread_self: 460651776
thread.c: 197:spdk_get_thread: *NOTICE*: pthread_self: 460651776
thread.c: 197:spdk_get_thread: *NOTICE*: pthread_self: 460651776
thread.c: 197:spdk_get_thread: *NOTICE*: pthread_self: 460651776
thread.c: 197:spdk_get_thread: *NOTICE*: pthread_self: 460651776
thread.c: 197:spdk_get_thread: *NOTICE*: pthread_self: 460651776
thread.c: 197:spdk_get_thread: *NOTICE*: pthread_self: 460651776
thread.c: 197:spdk_get_thread: *NOTICE*: pthread_self: 460651776
thread.c: 197:spdk_get_thread: *NOTICE*: pthread_self: 460651776
thread.c: 197:spdk_get_thread: *NOTICE*: pthread_self: 460651776
thread.c: 197:spdk_get_thread: *NOTICE*: pthread_self: 460651776
thread.c: 197:spdk_get_thread: *NOTICE*: pthread_self: 460651776
thread.c: 197:spdk_get_thread: *NOTICE*: pthread_self: 460651776
thread.c: 197:spdk_get_thread: *NOTICE*: pthread_self: 460651776
thread.c: 197:spdk_get_thread: *NOTICE*: pthread_self: 460651776
thread.c: 197:spdk_get_thread: *NOTICE*: pthread_self: 460651776
thread.c: 197:spdk_get_thread: *NOTICE*: pthread_self: 460651776
thread.c: 197:spdk_get_thread: *NOTICE*: pthread_self: 460651776
thread.c: 197:spdk_get_thread: *NOTICE*: pthread_self: 460651776
thread.c: 197:spdk_get_thread: *NOTICE*: pthread_self: 460651776
thread.c: 197:spdk_get_thread: *NOTICE*: pthread_self: 460651776
thread.c: 197:spdk_get_thread: *NOTICE*: pthread_self: 460651776
thread.c: 161:spdk_free_thread: *NOTICE*: pthread_self: 460651776
```

The code for my implementation is [here](https://github.com/utsaslab/rustfs/blob/1.0-dev-fileSystemToy-wrapper-future/examples/hello_nvme_bdev_rust_wrapper/src/main.rs) and I call the Rust Async I/O framework through `spdk_app_start` is [this line](https://github.com/utsaslab/rustfs/blob/23579ba55690195627cfe744f013c1aa9a07fee2/examples/hello_nvme_bdev_rust_wrapper/src/main.rs#L83).  `opts.start` is just a Rust wrapper around `spdk_app_start`.

Thanks!

**Comment by jimharris** [comment]
Thanks!  I missed one more instrumentation we need - in spdk_get_io_channel, just before the call to _get_thread().  That's the one that is failing.

**Comment by xxks-kkk** [comment]
Here is the updated output:

```
Enter main
Starting SPDK v18.07 / DPDK 18.05.0 initialization...
[ DPDK EAL parameters: hello_bdev -c 0x1 --legacy-mem --file-prefix=spdk_pid6004 ]
EAL: Detected 4 lcore(s)
EAL: Detected 1 NUMA nodes
EAL: Multi-process socket /var/run/dpdk/spdk_pid6004/mp_socket
EAL: No free hugepages reported in hugepages-1048576kB
EAL: Probing VFIO support...
EAL: VFIO support initialized
app.c: 530:spdk_app_start: *NOTICE*: Total cores available: 1
reactor.c: 718:spdk_reactors_init: *NOTICE*: Occupied cpu socket mask is 0x1
thread.c: 120:spdk_allocate_thread: *NOTICE*: pthread_self: 1881505024
reactor.c: 492:_spdk_reactor_run: *NOTICE*: Reactor started on core 0 on socket 0
thread.c: 197:spdk_get_thread: *NOTICE*: pthread_self: 1881505024
vbdev_passthru.c: 374:vbdev_passthru_init: *NOTICE*: conf parse matched: Malloc1
thread.c: 446:spdk_get_io_channel: *NOTICE*: pthread_self: 1881505024
thread.c: 446:spdk_get_io_channel: *NOTICE*: pthread_self: 1881505024
thread.c: 446:spdk_get_io_channel: *NOTICE*: pthread_self: 1881505024
thread.c: 446:spdk_get_io_channel: *NOTICE*: pthread_self: 1881505024
thread.c: 197:spdk_get_thread: *NOTICE*: pthread_self: 1881505024
thread.c: 446:spdk_get_io_channel: *NOTICE*: pthread_self: 1881505024
thread.c: 446:spdk_get_io_channel: *NOTICE*: pthread_self: 1881505024
vbdev_passthru.c: 478:vbdev_passthru_register: *NOTICE*: Match on Malloc1
vbdev_passthru.c: 512:vbdev_passthru_register: *NOTICE*: io_device created at: 0x0x565376b51f40
vbdev_passthru.c: 523:vbdev_passthru_register: *NOTICE*: bdev opened
vbdev_passthru.c: 534:vbdev_passthru_register: *NOTICE*: bdev claimed
thread.c: 446:spdk_get_io_channel: *NOTICE*: pthread_self: 1881505024
thread.c: 446:spdk_get_io_channel: *NOTICE*: pthread_self: 1881505024
thread.c: 446:spdk_get_io_channel: *NOTICE*: pthread_self: 1881505024
thread.c: 446:spdk_get_io_channel: *NOTICE*: pthread_self: 1881505024
thread.c: 446:spdk_get_io_channel: *NOTICE*: pthread_self: 1881505024
thread.c: 446:spdk_get_io_channel: *NOTICE*: pthread_self: 1881505024
thread.c: 197:spdk_get_thread: *NOTICE*: pthread_self: 1881505024
thread.c: 446:spdk_get_io_channel: *NOTICE*: pthread_self: 1881505024
thread.c: 446:spdk_get_io_channel: *NOTICE*: pthread_self: 1881505024
vbdev_passthru.c: 545:vbdev_passthru_register: *NOTICE*: pt_bdev registered
vbdev_passthru.c: 546:vbdev_passthru_register: *NOTICE*: created pt_bdev for: PT0
thread.c: 197:spdk_get_thread: *NOTICE*: pthread_self: 1881505024
EAL: PCI device 0000:02:00.0 on NUMA socket 0
EAL:   probe driver: 8086:953 spdk_nvme
EAL:   0000:02:00.0 VFIO group is not viable!
EAL: Requested device 0000:02:00.0 cannot be used
bdev_nvme.c:1388:bdev_nvme_library_init: *ERROR*: NVMe SSD "0000:00:04.0" could not be found.
bdev_nvme.c:1389:bdev_nvme_library_init: *ERROR*: Check PCIe BDF and that it is attached to UIO/VFIO driver.
bdev_nvme.c:1388:bdev_nvme_library_init: *ERROR*: NVMe SSD "0000:02:00.0" could not be found.
bdev_nvme.c:1389:bdev_nvme_library_init: *ERROR*: Check PCIe BDF and that it is attached to UIO/VFIO driver.
thread.c: 197:spdk_get_thread: *NOTICE*: pthread_self: 1881505024
thread.c: 197:spdk_get_thread: *NOTICE*: pthread_self: 1881505024
thread.c: 197:spdk_get_thread: *NOTICE*: pthread_self: 1881505024
thread.c: 446:spdk_get_io_channel: *NOTICE*: pthread_self: 1881505024
thread.c: 197:spdk_get_thread: *NOTICE*: pthread_self: 1881505024
thread.c: 197:spdk_get_thread: *NOTICE*: pthread_self: 1881505024
thread.c: 197:spdk_get_thread: *NOTICE*: pthread_self: 1881505024
thread.c: 197:spdk_get_thread: *NOTICE*: pthread_self: 1881505024
thread.c: 197:spdk_get_thread: *NOTICE*: pthread_self: 1881505024
thread.c: 197:spdk_get_thread: *NOTICE*: pthread_self: 1881505024
bdev name: Malloc0
bdev name: Malloc1
bdev name: PT0
Successfully open the device
thread.c: 446:spdk_get_io_channel: *NOTICE*: pthread_self: 1681192704
thread.c: 449:spdk_get_io_channel: *ERROR*: No thread allocated
Successful
thread.c: 197:spdk_get_thread: *NOTICE*: pthread_self: 1881505024
thread.c: 197:spdk_get_thread: *NOTICE*: pthread_self: 1881505024
thread.c: 197:spdk_get_thread: *NOTICE*: pthread_self: 1881505024
thread.c: 197:spdk_get_thread: *NOTICE*: pthread_self: 1881505024
thread.c: 197:spdk_get_thread: *NOTICE*: pthread_self: 1881505024
thread.c: 197:spdk_get_thread: *NOTICE*: pthread_self: 1881505024
thread.c: 197:spdk_get_thread: *NOTICE*: pthread_self: 1881505024
thread.c: 197:spdk_get_thread: *NOTICE*: pthread_self: 1881505024
thread.c: 197:spdk_get_thread: *NOTICE*: pthread_self: 1881505024
thread.c: 197:spdk_get_thread: *NOTICE*: pthread_self: 1881505024
thread.c: 197:spdk_get_thread: *NOTICE*: pthread_self: 1881505024
thread.c: 197:spdk_get_thread: *NOTICE*: pthread_self: 1881505024
thread.c: 197:spdk_get_thread: *NOTICE*: pthread_self: 1881505024
thread.c: 197:spdk_get_thread: *NOTICE*: pthread_self: 1881505024
thread.c: 197:spdk_get_thread: *NOTICE*: pthread_self: 1881505024
thread.c: 197:spdk_get_thread: *NOTICE*: pthread_self: 1881505024
thread.c: 197:spdk_get_thread: *NOTICE*: pthread_self: 1881505024
thread.c: 197:spdk_get_thread: *NOTICE*: pthread_self: 1881505024
thread.c: 197:spdk_get_thread: *NOTICE*: pthread_self: 1881505024
thread.c: 197:spdk_get_thread: *NOTICE*: pthread_self: 1881505024
thread.c: 197:spdk_get_thread: *NOTICE*: pthread_self: 1881505024
thread.c: 197:spdk_get_thread: *NOTICE*: pthread_self: 1881505024
thread.c: 197:spdk_get_thread: *NOTICE*: pthread_self: 1881505024
thread.c: 161:spdk_free_thread: *NOTICE*: pthread_self: 1881505024
```

**Comment by xxks-kkk** [comment]
I compare the output with `hello_bdev` output, I notice that the pthread id changes in the error printout (`1681192704`) while in `hello_bdev`, the same thread id is used when calling `spdk_get_io_channel` (i.e., one thread id across the lifetime of the program).

**Comment by jimharris** [comment]
I'm not very familiar with the Rust runtime.  But this definitely confirms that these async fns are being executed from a different pthread context.

**Comment by xxks-kkk** [comment]
Hello Jim,

I take a look at the fio example in spdk. I'm not sure I fully understand the example. I know all the read/write request is sent via io_channel. Once I allocate a io_channel using `spdk_allocate_thread`, I'm not sure how can I reference it to send the requests. Can I use `spdk_bdev_get_io_channel` to immediately reference the channel I just allocated?

Thanks!

**Comment by jimharris** [comment]
Hi @xxks-kkk,

spdk_allocate_thread() alllocates an spdk_thread object - not an spdk_io_channel.  You just need to call spdk_allocate_thread() before any subsequent spdk_bdev_get_io_channel() calls.

Think of an io_channel as a logical entity that provides the ability to send an I/O to one device on one thread.  The io_channel may only be used on the thread from which it was originally allocated.  If another caller requests an io_channel for that device on the same thread, it will get a reference to the first one that was allocated.  So SPDK needs to track the io_channels per device and per thread.

Take logical volumes for example.  There may be 100 logical volumes in a logical volume store on an NVMe namespace.  This could result in an upper layer (iSCSI, NVMe-oF, vhost) requesting an io_channel for each of these logical volumes.  Since each of the logical volumes is distinct, this will result in the creation of 100 different io_channels.  For now, let's just imagine that the upper layer is requesting all of those io_channels on one thread.

But each of those logical volumes then needs an io_channel to send I/O to the underlying NVMe namespace!  But since this is all happening on one thread, we don't need 100 different io_channels for the NVMe namespace - this would result in creating 100 different NVMe queue pairs, which would be a waste (few SSDs support that many queue pairs).  So the spdk_get_io_channel() functionality is smart enough to know that they are all requesting an io_channel for the same io_device (the NVMe namespace), and will return the same io_channel to each of the callers.

So back to the original question - we need to track the threads that are participating in this framework within SPDK.  Each thread must call spdk_allocate_thread() before it can get io_channels.  That thread is also responsible for calling spdk_free_thread() once it is done.  Typically this is done at thread creation/destruction and not an as-needed basis.

-Jim

**Comment by benlwalker** [comment]
SPDK is also going through a bit of a transition at the moment regarding this threading stuff, and we're doing it specifically to allow it to plug into futures/promises frameworks in a much more straightforward way. One of the main frameworks on my radar is Rust's tokio, so I'm glad you're looking at it!

SPDK is fundamentally designed as a cooperative multi-tasking system. All of the libraries consist of functions that do some small amount of work and then immediately return ("yield"), without ever blocking. They're also designed to coordinate across threads via message passing (again, without blocking) instead of taking locks.

SPDK assumes the host application consists of some number of system threads, and for each system thread today we need to initiate it as an spdk_thread. This sets up the queues to pass messages essentially. You do this by calling spdk_allocate_thread() on the system thread when it first starts up. You can only make SPDK calls from system threads where spdk_allocate_thread() has been previously called (with a few exceptions, such as the NVMe driver which is lower level and makes no assumptions about anything).

SPDK provides an application framework (lib/event) that will spawn one system thread per core and automatically call spdk_allocate_thread() on each one. It will also handle processing incoming messages and pollers on that system thread - it owns the event loop/reactor on that system thread. You don't have to use the SPDK event framework to use SPDK though. You can call spdk_allocate_thread() directly and handle the incoming messages and pollers on your own. This is the part that we're currently clarifying and simplifying, but the bdev fio_plugin is the current best example of how to do this.

So basically, for any new system thread you spawn in rust, you need to call spdk_allocate_thread() and then enter an event loop to process incoming messages and pollers on that thread. Fortunately for you, rust has effectively no run time. That means it isn't spawning any threads on your behalf and you have full control over when threads are spawned and can insert the calls as necessary. The easiest way to get everything incorporated right now is to not spawn any threads from rust, if you can find a way to make it do that.

**Comment by gila** [comment]
@benlwalker nice to see you looking into that -- we have been using rust with SPDK very successfully, and we actually use it as a higher abstraction to create more bdevs specific to our use case. We actually started out with reimplementing the null bdev and got a surprising, rather significant, performance improvements. The code is not the exact same ofcourse due to the nature of rust, and I only added support for a single core as we use it mostly as a tool to measure overhead per vbdev we add. But -- having all that said, I was surprised to see that. 

In any case, there are some other things that make the use of rust... let's say unfriendly. For example, registering new bdev modules is currently done by macro's which we can easily work around but maybe, as low priority backlog item -- it might be useful to discuss and make it less painfull. We could do these things during a hackathon or discuss at an SPDK conference for example as they are just annoying at best. 

@xxks-kkk your are most certainly creating threads as you are using the runtime as is, and you can also see that from the debug output @jimharris asked for. You could follow @benlwalker advise but it's not really going to help you (other then it might work). Rather you need to create an executor such that the futures get scheduled in such a way it makes sense from an SPDK standpoint only then you can get a meaningful abstraction from a tokio perspective. Now having that said, it works just fine without, and depending on you want to do its not all that bad in terms of callbacks. If you want to consume the blob related stuff (where there are a lot more of callback sequences) then yes it might be worth to do.

**Comment by jimharris** [comment]
@gila - would you be able to describe your workaround for the bdev constructor macros?  For your use case, are any of the other constructor macro types causing problems for integrating into Rust?

I know I'd certainly be interested in seeing how we can get upstream SPDK to make it easier to integrate SPDK into Rust applications.  I love the idea of talking about this at a future hackathon but maybe it's something we can start talking about sooner than that!  :)

-Jim

**Comment by xxks-kkk** [comment]
@gila I sense you talk about [Starfish](https://github.com/jkozlowski/starfish) repo, which I have been watching for quite some time now. Overall, I like the future style over the callbacks (SPDK official doc states that lack of futures is one of the drawback they have to workaround in C). However, my approach now is to avoid re-invent wheels whenever possible (i.e., use Tokio as much as I can). I'm not sure what you mean by

> you need to create an executor such that the futures get scheduled in such a way it makes sense from an SPDK standpoint only then you can get a meaningful abstraction from a tokio perspective

I'm consistently get confused by the concepts of poller, executor, and event functions both mentioned in Tokio and SPDK. I'm wondering if you can explain more about the sentence I quote above? Also, I'm wondering if you know any background story on why Starfish decides to write its own executor instead of using something from Tokio?

@benlwalker Glad tokio is on your radar because that's exactly what I have been trying to wrap SPDK around with Rust.  I'm wondering if support futures framework in SPDK is something that is being planned for the next release?

**Comment by gila** [comment]
@jimharris 

The macro's in C, can not directly be called by RUST. So that's one "problem". But the biggest issue is not really related to macro's themselves but more of a linker problem when creating a static binary -- that requires us to reference (but not call) functions in SPDK (and DPDK for that matter) to prevent the linker to remove them during the final linking stage. 

Rust does not (yet?) give good control over these things. The issue then with the `SPDK_BDEV_MODULE_REGISTER()` in particular is it cannot be used outside of the file (static void) and it uses a line number to generate the actual function names and thus may change from release to release. 

Also -- functions that are inlined are not callable (i.e spdk_io_channel_get_ctx()). And sometimes its a little annoying that certain structures enforce linked lists (e.g bdev aliases) and there are no public interfaces for this to "walk" them. Lastly, it would be nice if could reduce the number of void pointers to get a more clear interface. So as mentioned, all in all, no real big issues IMHO. 

From an interface perspective, the RPC/JSON API is the most difficult to use as the abstractions provided does not allow for custom serializers/deserializers.


@xxks-kkk 
Our use case is different compared to the project you are referring to but in general, the Tokio framework is built around the notion of using futures. As for the confusion -- more than willing to chat with you about things but not through the SPDK issue tracker :)
